### PR TITLE
Upgrade to XGBoost 1.2.0

### PR DESCRIPTION
**Pull request by podcastinator**
Main change: Upgrade backend to XGBoost 1.2.0

Other changes:
1. Add APIs that do not require consensus across clients, and a C demo using this API
2. Move around files to mimic XGBoost file organization
3. Improve names of assorted key files

### CODE REVIEW

File: include/xgboost/c_api.h [hunk 1]
```diff
@@ -145,55 +108,27 @@ XGB_DLL int XGBCreateEnclave(const char *enclave_image, int log_verbosity);
 /*!
  * \brief load a data matrix
  * \param fname the name of the file
+ * \param username the name of the file owner
  * \param silent whether print messages during loading
  * \param out a loaded data matrix
  * \return 0 when success, -1 when failure happens
  */
 XGB_DLL int XGDMatrixCreateFromFile(const char *fname,
+                                    char* username,
```
Comment (chester-leung) [inline include/xgboost/c_api.h:117]: Do we now support loading from unencrypted files?

File: python-package/securexgboost/core.py [hunk 2]
```diff
@@ -1623,6 +1623,16 @@ def predict(self, data, output_margin=False, ntree_limit=0, pred_leaf=False,
             pred_contribs), and the sum of the entire matrix equals the raw untransformed margin
             value of the prediction. Note the last row and column correspond to the bias term.
 
+        training : bool
+            Whether the prediction value is used for training.  This can effect
```
Comment (chester-leung) [inline python-package/securexgboost/core.py:1627]: What does this mean? Does this change affect us?

File: config/gen-client.sh [hunk 3]
```diff
@@ -15,7 +15,7 @@ openssl req -new -key ${USERNAME}.pem -out ${USERNAME}.csr -subj "/CN=${USERNAME
 
 # Generate a certificate for the client signed by the signer
```
Comment (chester-leung) [inline config/gen-client.sh]: Let's change this comment to "Generate a certificate for the client signed by the root CA" to reflect the file name change?

File: demo/c-api/c-api-demo.cc [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+#define safe_xgboost(call) {                                            \
+int err = (call);                                                       \
+if (err != 0) {                                                         \
+  fprintf(stderr, "%s:%d: error in %s: %s\n", __FILE__, __LINE__, #call, XGBGetLastError()); \
+  exit(1);                                                              \
+}                                                                       \
+}
+
+int main(int argc, char** argv) {
```
Comment (chester-leung) [inline demo/c-api/c-api-demo.cc:36]: Should we add a note saying that the C API demo will only work if building the library with consensus turned off? Maybe in this file as well as `demo/c-api/c-api-demo.cc`

File: enclave/dmlc-core/src/data.cc [hunk 5]
```diff
@@ -94,34 +95,91 @@ CreateParser_(const char *uri_,
...
+ *  }
+ *}
  */
 
 DMLC_REGISTER_PARAMETER(LibSVMParserParam);
 DMLC_REGISTER_PARAMETER(LibFMParserParam);
 DMLC_REGISTER_PARAMETER(CSVParserParam);
 }  // namespace data
 
+// template specialization
```
[the remark this answers is not in the document]
Comment (podcastinator) [reply enclave/dmlc-core/src/data.cc:126]: Yeah

File: include/xgboost/c_api.h (hunk 1 above)
Comment (podcastinator) [reply include/xgboost/c_api.h:117]: Yes!

File: python-package/securexgboost/core.py (hunk 2 above)
Comment (podcastinator) [reply python-package/securexgboost/core.py:1627]: It does in the sense that I had to modify all the APIs to incorporate it.
Essentially it seems to affect the behavior of training if the Booster is of type `dart` (it is `gbtree` by default).

File: demo/c-api/c-api-demo.cc (hunk 4 above)
Comment (podcastinator) [reply demo/c-api/c-api-demo.cc:36]: Which other file are you referring to?

Comment (chester-leung) [reply demo/c-api/c-api-demo.cc:36]: Oh oops I referred to the same file 😅 I meant `demo/c-api/README.md`

File: include/xgboost/c_api.h (hunk 1 above)
Comment (chester-leung) [reply include/xgboost/c_api.h:117]: Awesome!

Comment (podcastinator) [reply include/xgboost/c_api.h:117]: Although, the Python API still needs to be fixed (along with the RPC stuff).

File: demo/c-api/c-api-demo.cc (hunk 4 above)
Comment (podcastinator) [reply demo/c-api/c-api-demo.cc:36]: Ah lol ok

**Review by chester-leung — APPROVED** [review]

**Pull request merged**